sansa-rdf-spark/src/main/scala/net/sansa_stack/rdf/spark/qualityassessment/metrics/availability/DereferenceableUris.scala

```diff
@@ -1,72 +1,80 @@
 package net.sansa_stack.rdf.spark.qualityassessment.metrics.availability
 
-import org.apache.spark.rdd.RDD
-import org.apache.jena.graph.{ Triple, Node }
 import net.sansa_stack.rdf.spark.qualityassessment.utils.NodeUtils._
+import org.apache.jena.graph.{ Node, Triple }
+import org.apache.spark.rdd.RDD
 
 /*
  * Dereferenceability of the URI.
  * @author Gezim Sejdiu
  */
 object DereferenceableUris {
-  implicit class DereferenceableUrisFunctions(dataset: RDD[Triple]) extends Serializable {
+
+  /**
+   * This metric calculates the number of valid redirects of URI.
+   * It computes the ratio between the number of all valid redirects
+   * (subject + predicates + objects)a.k.a dereferencedURIS and
+   * the total number of URIs on the dataset.
+   */
+  def assessDereferenceableUris(dataset: RDD[Triple]): Double = {
 
     val totalURIs = dataset.filter(_.getSubject.isURI())
       .union(dataset.filter(_.getPredicate.isURI()))
       .union(dataset.filter(_.getObject.isURI()))
-      .distinct().cache()
-
-    val totalSubjects = dataset.filter(_.getSubject.isURI()).cache()
-    val totalObjects = dataset.filter(_.getObject.isURI()).cache()
+      .distinct()
 
     // check object if URI and local
     val objects = dataset.filter(f =>
-      f.getObject.isURI() && isInternal(f.getObject) && !isBroken(f.getObject)).cache()
+      f.getObject.isURI() && isInternal(f.getObject) && !isBroken(f.getObject))
```

Comment:
1. You should use better variable names, e.g. `f` is confusing given that the dataset contains triples something like `t` might help reading/understanding
2. Could be more concise given that you're doing the same things several times here. For example:

``` scala
def filterLocalNodes(ds: RDD[Triple], f: Triple => Node): RDD[Triple] = ds.filter(t => isLocalURI(f(t)))
def isLocalURI(node: Node): Boolean = node.isURI() && isInternal(node) && !isBroken(node)

val predicates = filterLocalNodes(dataset, _.getPredicate)
```
Not shorter, but I guess you know what I mean :smiley:

Reply:
@LorenzBuehmann many thanks for your valuable feedback! Indeed it requires a refactoring this part and will get it done as soon as I will be able to jump to this part again!
Thanks once more.